README.md
# cdk8s-redis-sts  
![Release](https://github.com/opencdk8s/cdk8s-redis-sts/workflows/Release/badge.svg?branch=development)
[![npm version](https://badge.fury.io/js/%40opencdk8s%2Fcdk8s-redis-sts.svg)](https://badge.fury.io/js/%40opencdk8s%2Fcdk8s-redis-sts)
[![PyPI version](https://badge.fury.io/py/cdk8s-redis-sts.svg)](https://badge.fury.io/py/cdk8s-redis-sts)
![npm](https://img.shields.io/npm/dt/@opencdk8s/cdk8s-redis-sts?label=npm&color=green) 
![PyPi](https://img.shields.io/pypi/dm/cdk8s-redis-sts?label=pypi&color=green) 

Create a Replicated Redis Statefulset on Kubernetes, powered by the [cdk8s project](https://cdk8s.io) 🚀

## Disclaimer 

This construct is under heavy development, and breaking changes will be introduced very often. Please don't forget to version lock your code if you are using this construct.

## Overview

**cdk8s-redis-sts** is a [cdk8s](https://cdk8s.io) library.

```typescript
import { Construct } from 'constructs';
import { App, Chart, ChartProps } from 'cdk8s';
import { MyRedis } from '@opencdk8s/cdk8s-redis-sts';

export class MyChart extends Chart {
  constructor(scope: Construct, id: string, props: ChartProps = { }) {
    super(scope, id, props);
    new MyRedis(this, 'dev', {
        image: 'redis',
        namespace: 'databases',
        volumeSize: '10Gi',
        replicas: 2,
        createStorageClass: true,
        volumeProvisioner: 'kubernetes.io/aws-ebs',
        storageClassName: "io1-slow",
        storageClassParams: {
          type: 'io1',
          fsType: 'ext4',
          iopsPerGB: "10",
        },
    });
    }
}

const app = new App();
new MyChart(app, 'dev');
app.synth();
```

Then the Kubernetes manifests created by `cdk8s synth` command will have Kubernetes resources such as `Statefulset`, `ConfigMap` and `Service` as follows.

<details>
<summary>manifest.k8s.yaml</summary>

```yaml
allowVolumeExpansion: true
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: io1-slow
parameters:
  fsType: ext4
  iopsPerGB: "10"
  type: io1
provisioner: kubernetes.io/aws-ebs
reclaimPolicy: Retain
---
apiVersion: v1
data:
  master.conf: |-
    
    bind 0.0.0.0
    daemonize no
    port 6379
    tcp-backlog 511
    timeout 0
    tcp-keepalive 300
    supervised no
  slave.conf: |-
    
    slaveof dev 6379
kind: ConfigMap
metadata:
  name: dev-redis-conf
---
apiVersion: v1
kind: Service
metadata:
  labels:
    app: dev
  name: dev
  namespace: databases
spec:
  ports:
    - port: 6379
      targetPort: 6379
  selector:
    app: dev
  type: ClusterIP
---
apiVersion: apps/v1
kind: StatefulSet
metadata:
  labels:
    app: dev
  name: dev
  namespace: databases
spec:
  replicas: 2
  selector:
    matchLabels:
      app: dev
  serviceName: dev
  template:
    metadata:
      labels:
        app: dev
    spec:
      containers:
        - command:
            - bash
            - -c
            - |-
              [[ `hostname` =~ -([0-9]+)$ ]] || exit 1
              ordinal=${BASH_REMATCH[1]}
              if [[ $ordinal -eq 0 ]]; then
              redis-server /mnt/redis/master.conf
              else
              redis-server /mnt/redis/slave.conf
              fi
          env: []
          image: redis
          name: redis
          ports:
            - containerPort: 6379
          resources:
            limits:
              cpu: 400m
              memory: 512Mi
            requests:
              cpu: 200m
              memory: 256Mi
          volumeMounts:
            - mountPath: /data
              name: dev
            - mountPath: /mnt/redis/
              name: dev-redis-conf
      terminationGracePeriodSeconds: 10
      volumes:
        - configMap:
            name: dev-redis-conf
          name: dev-redis-conf
  volumeClaimTemplates:
    - metadata:
        name: dev
        namespace: databases
      spec:
        accessModes:
          - ReadWriteOnce
        resources:
          requests:
            storage: 10Gi
        storageClassName: io1-slow
```

</details>

## Installation

### TypeScript

Use `yarn` or `npm` to install.

```sh
$ npm install @opencdk8s/cdk8s-redis-sts
```

```sh
$ yarn add @opencdk8s/cdk8s-redis-sts
```

### Python

```sh
$ pip install cdk8s-redis-sts
```

## Contribution

1. Fork ([link](https://github.com/opencdk8s/cdk8s-redis-sts/fork))
2. Bootstrap the repo:
  
    ```bash
    npx projen   # generates package.json 
    yarn install # installs dependencies
    ```
3. Development scripts:
   |Command|Description
   |-|-
   |`yarn compile`|Compiles typescript => javascript
   |`yarn watch`|Watch & compile
   |`yarn test`|Run unit test & linter through jest
   |`yarn test -u`|Update jest snapshots
   |`yarn run package`|Creates a `dist` with packages for all languages.
   |`yarn build`|Compile + test + package
   |`yarn bump`|Bump version (with changelog) based on [conventional commits]
   |`yarn release`|Bump + push to `master`
4. Create a feature branch
5. Commit your changes
6. Rebase your local changes against the master branch
7. Create a new Pull Request (use [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) for the title please)

## Licence

[Apache License, Version 2.0](./LICENSE)

## Author

[Hunter-Thompson](https://github.com/Hunter-Thompson)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.7](https://github.com/opencdk8s/cdk8s-redis-sts/compare/v0.0.6...v0.0.7) (2021-02-15)


### Features

* **cm:** create configmap if not specified ([#6](https://github.com/opencdk8s/cdk8s-redis-sts/issues/6)) ([1fedde6](https://github.com/opencdk8s/cdk8s-redis-sts/commit/1fedde683cf7fde8cf4a742127352654f9bad1e4))

### [0.0.6](https://github.com/opencdk8s/cdk8s-redis-sts/compare/v0.0.5...v0.0.6) (2021-02-13)


### Features

* option to create storage class ([#5](https://github.com/opencdk8s/cdk8s-redis-sts/issues/5)) ([cbab1cc](https://github.com/opencdk8s/cdk8s-redis-sts/commit/cbab1cc2b5b6776bbcfb68fd0b31221733c65688))

### [0.0.5](https://github.com/opencdk8s/cdk8s-redis-sts/compare/v0.0.4...v0.0.5) (2021-02-11)


### Bug Fixes

* revert last commit ([0713622](https://github.com/opencdk8s/cdk8s-redis-sts/commit/07136222f27b48da03e2a26ad6079fc7638cab59))

### [0.0.4](https://github.com/opencdk8s/cdk8s-redis-sts/compare/v0.0.3...v0.0.4) (2021-02-11)

### [0.0.3](https://github.com/Hunter-Thompson/cdk8s-redis-sts/compare/v0.0.2...v0.0.3) (2021-02-10)

### [0.0.2](https://github.com/Hunter-Thompson/cdk8s-redis-sts/compare/v0.0.1...v0.0.2) (2021-02-10)

### 0.0.1 (2021-02-10)


.projenrc.js
const { ConstructLibraryCdk8s } = require('projen');

const project = new ConstructLibraryCdk8s({
  author: 'Hunter Thompson',
  authorAddress: 'aatman@auroville.org.in',
  cdk8sVersion: '1.0.0-beta.8',
  defaultReleaseBranch: 'development',
  stability: 'experimental',
  jsiiFqn: 'projen.ConstructLibraryCdk8s',
  name: '@opencdk8s/cdk8s-redis-sts',
  npmAccess: 'public',
  repositoryUrl: 'https://github.com/opencdk8s/cdk8s-redis-sts',

  python: {
    distName: 'cdk8s-redis-sts',
    module: 'cdk8s_redis_sts',
  },
  peerDeps: ['constructs@^3.3.5'],
  releaseEveryCommit: false,
  devDeps: [
    'constructs@^3.3.5',
    'prettier@^2.2.1',
    'jsii-pacmak@^1.20.1',
    '@commitlint/config-conventional@^11.0.0',
    '@commitlint/cli@^11.0.0',
    'eslint-import-resolver-node@^0.3.4',
    'eslint-import-resolver-typescript@^2.3.0',
    'eslint-plugin-import@^2.22.1',
    'eslint@^7.19.0',
    'jsii-diff@^1.20.1',
    'jsii-docgen@^1.8.36',
    'jsii@^1.20.1',
    'json-schema@^0.3.0',
  ],

  dependabot: false,
  gitignore: ['package.json', 'test/'],
  pullRequestTemplate: false,
  releaseBranches: ['development'],
  codeCov: true,
  clobber: false,
  readme: true,
});

const common_exclude = ['cdk.out', 'package.json', 'yarn-error.log', 'coverage', '.DS_Store', '.idea', '.vs_code'];
project.gitignore.exclude(...common_exclude);

project.synth();


src/index.ts
import { Construct } from 'constructs';
import * as k8s from './imports/k8s';


export interface StsOpts {
  /**
   * @default 1
   */
  readonly replicas?: number;

  /**
   * Container image.
   */
  readonly image: string;

  /**
   * namespace
   * @default - default
   */
  readonly namespace: string;

  /**
   * Option to create storage class, if enabled, a storage class will be created for the statefulset
   * @default true
   */
  readonly createStorageClass?: boolean;

  /**
   * The storage class to use for our PVC
   * @default 'gp2-expandable'
   */
  readonly storageClassName?: string;

  /**
   * The name of the configmap to use instead of the default one
   */
  readonly configMapName?: string;

  /**
   * Each StorageClass has a provisioner that determines what volume plugin is used for provisioning PVs. This field must be specified.
   * See [this](https://kubernetes.io/docs/concepts/storage/storage-classes/#provisioner) for Ref
   * @default 'kubernetes.io/aws-ebs'
   */
  readonly volumeProvisioner?: string;

  /**
   * Storage class params
   * @default - { type = gp2, fsType: ext4 }
   */
  readonly storageClassParams?: { [name: string]: string };

  /**
   * nodeSelector params
   * @default - undefined
   */
  readonly nodeSelectorParams?: { [name: string]: string };

  /**
   * Additional labels to apply to resources.
   * @default - none
   */
  readonly labels?: { [name: string]: string };

  /**
   * Resources requests for the DB.
   * @default - Requests = { CPU = 200m, Mem = 256Mi }, Limits = { CPU = 400m, Mem = 512Mi }
   */
  readonly resources?: ResourceRequirements;


  /**
   * The Volume size of our DB in string, e.g 10Gi, 20Gi
   */
  readonly volumeSize?: string;

  /**
   * Environment variables to pass to the pod
   */
  readonly env?: { [name: string]: string };

}

export interface ResourceRequirements {
  /**
   * Maximum resources for the web app.
   * @default - CPU = 400m, Mem = 512Mi
   */
  readonly limits?: ResourceQuantity;

  /**
   * Required resources for the web app.
   * @default - CPU = 200m, Mem = 256Mi
   */
  readonly requests?: ResourceQuantity;
}

export interface ResourceQuantity {
  /**
   * @default - no limit
   */
  readonly cpu?: string;

  /**
   * @default - no limit
   */
  readonly memory?: string;
}


export class MyRedis extends Construct {
  public readonly name: string;
  public readonly namespace: string;


  constructor(scope: Construct, name: string, opts: StsOpts) {
    super(scope, name);

    const namespace = opts.namespace ?? 'default';
    this.namespace = namespace;
    var configMapName = opts.configMapName ?? name;
    var storageClassName = opts.storageClassName ?? 'gp2';
    const volumeProvisioner = opts.volumeProvisioner ?? 'kubernetes.io/aws-ebs';
    const storageClassParams = opts.storageClassParams ?? { type: 'gp2', fsType: 'ext4' };
    const volumeRequest = {
      storage: k8s.Quantity.fromString(String(opts.volumeSize)),
    };
    const replicas = opts.replicas ?? 1;
    const resources = {
      limits: convertQuantity(opts.resources?.limits, {
        cpu: '400m',
        memory: '512Mi',
      }),
      requests: convertQuantity(opts.resources?.requests, {
        cpu: '200m',
        memory: '256Mi',
      }),
    };
    const nodeSelectorParams = opts.nodeSelectorParams ?? undefined;


    const label = {
      ...opts.labels,
      app: name,
    };

    if (opts.createStorageClass === true) {
      const storageClassOpts: k8s.KubeStorageClassProps = {
        metadata: {
          name: storageClassName,
        },
        provisioner: volumeProvisioner,
        allowVolumeExpansion: true,
        reclaimPolicy: 'Retain',
        parameters: {
          ...storageClassParams,
        },
      };
      const storageclass = new k8s.KubeStorageClass(this, 'storageclass', storageClassOpts);
      this.name = storageclass.name;
      var storageClassName = storageclass.name;
    }

    if (configMapName != name) {
      // pass
    } else {
      const configMapOpts: k8s.KubeConfigMapProps = {
        metadata: {
          name: `${name}-redis-conf`,
          namespace: namespace,
        },
        data: {
          'master.conf':
          '\nbind 0.0.0.0' +
          '\ndaemonize no' +
          '\nport 6379' +
          '\ntcp-backlog 511' +
          '\ntimeout 0' +
          '\ntcp-keepalive 300' +
          '\nsupervised no',
          'slave.conf': `\nslaveof ${name} 6379`,
        },
      };
      const configmap = new k8s.KubeConfigMap(this, 'cm', configMapOpts);
      this.name = configmap.name;
      var configMapName = configmap.name;
    }

    const serviceOpts: k8s.KubeServiceProps = {
      metadata: {
        labels: label,
        name: name,
        namespace: namespace,
      },
      spec: {
        type: 'ClusterIP',
        selector: label,
        ports: [{ port: 6379, targetPort: k8s.IntOrString.fromNumber(6379) }],
      },
    };

    const svc = new k8s.KubeService(this, 'service', serviceOpts);
    this.name = svc.name;

    const pvcProps: k8s.KubePersistentVolumeClaimProps = {
      metadata: {
        name: name,
        namespace: namespace,
      },
      spec: {
        accessModes: ['ReadWriteOnce'],
        storageClassName: storageClassName,
        resources: {
          requests: volumeRequest,
        },
      },
    };

    // const initRedis: k8s.Container = {
    //   name: 'init-redis',
    //   image: opts.image,
    //   command: [
    //     'bash',
    //     '-c',
    //     'set -ex\n# Generate mysql server-id from pod ordinal index.\nmkdir -p /etc/redis\n[[ `hostname` =~ -([0-9]+)$ ]] || exit 1\nordinal=${BASH_REMATCH[1]}\n# Copy appropriate conf.d files from config-map to emptyDir.\nif [[ $ordinal -eq 0 ]]; then\ncp /mnt/redis/master.conf /etc/redis/redis.conf\nelse\ncp /mnt/redis/slave.conf /etc/redis/redis.conf\nfi',
    //   ],
    //   volumeMounts: [{
    //     name: name,
    //     mountPath: '/data',
    //   },
    //   {
    //     name: `${name}-redis-conf`,
    //     mountPath: '/mnt/config-map',
    //   }],
    // };

    const redis: k8s.Container = {
      name: 'redis',
      image: opts.image,
      ports: [{
        containerPort: 6379,
      }],
      command: [
        'bash',
        '-c',
        '[[ `hostname` =~ -([0-9]+)$ ]] || exit 1\nordinal=${BASH_REMATCH[1]}\nif [[ $ordinal -eq 0 ]]; then\nredis-server /mnt/redis/master.conf\nelse\nredis-server /mnt/redis/slave.conf\nfi',
      ],
      resources: resources,
      env: renderEnv(opts.env),
      volumeMounts: [{
        name: name,
        mountPath: '/data',
      },
      {
        name: `${name}-redis-conf`,
        mountPath: '/mnt/redis/',
      }],
    };

    const stsOpts: k8s.KubeStatefulSetProps = {
      metadata: {
        labels: label,
        name: name,
        namespace: namespace,
      },
      spec: {
        serviceName: svc.name,
        replicas: replicas,
        selector: { matchLabels: label },
        template: {
          metadata: { labels: label },
          spec: {
            containers: [
              redis,
            ],
            terminationGracePeriodSeconds: 10,
            nodeSelector: nodeSelectorParams,
            volumes: [{
              name: `${name}-redis-conf`,
              configMap: {
                name: `${configMapName}`,
              },
            }],
          },
        },
        volumeClaimTemplates: [pvcProps],
      },
    };
    const sts = new k8s.KubeStatefulSet(this, 'statefulset', stsOpts);
    this.name = sts.name;
  }
}

function renderEnv(env: { [key: string]: string } = { }): k8s.EnvVar[] {
  const result = new Array<k8s.EnvVar>();
  for (const [key, value] of Object.entries(env)) {
    result.push({
      name: key,
      value: value,
    });
  }
  return result;
}


/**
 * Converts a `ResourceQuantity` type to a k8s.Quantity map.
 *
 * If `user` is defined, the values provided there (or lack thereof) will be
 * passed on. This means that if the user, for example, did not specify a value
 * for `cpu`, this value will be omitted from the resource requirements. This is
 * intentional, in case the user intentionally wants to omit a constraint.
 *
 * If `user` is not defined, `defaults` are used.
 */

function convertQuantity(
  user: ResourceQuantity | undefined,
  defaults: { cpu: string; memory: string },
): { [key: string]: k8s.Quantity } {
  // defaults
  if (!user) {
    return {
      cpu: k8s.Quantity.fromString(defaults.cpu),
      memory: k8s.Quantity.fromString(defaults.memory),
    };
  }

  const result: { [key: string]: k8s.Quantity } = {};

  if (user.cpu) {
    result.cpu = k8s.Quantity.fromString(user.cpu);
  }

  if (user.memory) {
    result.memory = k8s.Quantity.fromString(user.memory);
  }

  return result;
}